File: 2023/14/14.rb
#!/usr/bin/env ruby

UNMOVABLE = '#'
MOVABLE = 'O'
EMPTY = '.'

def print_grid(grid)
  grid.each do |row|
    puts row.join('')
  end
  puts
end

def calculate_weight(grid)
  grid.each_with_index.map do |row, i|
    row.count { |c| c == MOVABLE } * (grid.size - i)
  end.sum
end

def roll(grid)
  new_grid = grid.map(&:dup)

  grid.each_with_index do |row, y|
    row.each_with_index do |chr, x|
      next unless chr == MOVABLE
      new_y = (y - 1).downto(0).find do |ny|
        new_grid[ny][x] != EMPTY
      end || -1 # Reached edge
      new_y += 1
      if new_y < y
        new_grid[new_y][x] = MOVABLE
        new_grid[y][x] = EMPTY
      end
    end
  end

  new_grid
end

def rotate_clockwise(grid)
  grid.transpose.map(&:reverse)
end

def cycle(grid)
  4.times do # north, west, south, east
    grid = roll(grid)
    grid = rotate_clockwise(grid)
  end
  grid
end

grid = []
$stdin.readlines.lazy.map(&:chomp).each do |line|
  grid << line.chars
end

ORIGINAL = grid.map(&:dup)

# Part 1
grid = roll(ORIGINAL)
puts calculate_weight(grid)

# Part 2
grid = ORIGINAL
CYCLE_COUNT = 1_000_000_000
history = [grid]
cycle_begin = 0

CYCLE_COUNT.times do |i|
  grid = cycle(grid)

  existing = history.index(grid)
  if existing
    cycle_begin = existing
    break
  end

  history << grid
end

# Find grid at CYCLE_COUNT based on cycle history
grid = history[cycle_begin + (CYCLE_COUNT - cycle_begin) % (history.size - cycle_begin)]
puts calculate_weight(grid)

File: 2023/14/14.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strings"
)

var MOVABLE byte = 'O'
var UNMOVABLE byte = '#'
var EMPTY byte = '.'

func ParseInput() []string {
	scanner := bufio.NewScanner(os.Stdin)
	grid := []string{}
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}
	return grid
}

func calculateWeight(grid []string) int {
	sum := 0
	for i, row := range grid {
		sum += strings.Count(row, string(MOVABLE)) * (len(grid) - i)
	}
	return sum
}

func rotateClockwise(grid []string) []string {
	newGrid := make([]string, len(grid[0]))
	for i := 0; i < len(grid[0]); i++ {
		row := make([]byte, len(grid))
		for j := 0; j < len(grid); j++ {
			row[len(grid)-j-1] = grid[j][i]
		}
		newGrid[i] = string(row)
	}
	return newGrid
}

func roll(grid []string) []string {
	newGrid := slices.Clone(grid)

	for y, row := range grid {
		for x, chr := range row {
			if byte(chr) != MOVABLE {
				continue
			}
			var ny int
			for ny = y - 1; ny >= 0; ny-- {
				if newGrid[ny][x] != EMPTY {
					break
				}
			}
			ny++
			if ny < y {
				newGrid[ny] = newGrid[ny][:x] + string(MOVABLE) + newGrid[ny][x+1:]
				newGrid[y] = newGrid[y][:x] + string(EMPTY) + newGrid[y][x+1:]
			}
		}
	}
	return newGrid
}

func cycle(grid []string) []string {
	grid = rotateClockwise(roll(grid))
	grid = rotateClockwise(roll(grid))
	grid = rotateClockwise(roll(grid))
	return rotateClockwise(roll(grid))
}

func main() {
	grid := ParseInput()
	grid = roll(grid)
	fmt.Println(calculateWeight(grid))

	history := [][]string{grid}
	cycle_count := 1000000000
	var cycle_begin int

	for i := 0; i <= cycle_count; i++ {
		grid = cycle(grid)

		cycle_begin = slices.IndexFunc(history, func(e []string) bool {
			for i, row := range e {
				if row != grid[i] {
					return false
				}
			}
			return true
		})

		if cycle_begin > -1 {
			break
		}
		history = append(history, grid)
	}

	grid = history[cycle_begin+(cycle_count-cycle_begin)%(len(history)-cycle_begin)]
	fmt.Println(calculateWeight(grid))
}
